### ISSUE TITLE
Case insensitive completion

### ISSUE THREAD

**Issue by hanche** [elves/elvish#893]
Given this setup:

```
edit:completion:matcher[argument] = [seed]{ edit:match-prefix $seed &ignore-case=$true }
```

If I try to complete a filename and there is only one match, but the case of what I typed does not match the filename, I get a completion list containing just one element. For example, after typing `ls doc` (all lower case) and hitting TAB, I get this (`⬥` is my prompt. I removed the right prompt)
```
⬥ ls Documents/
 COMPLETING argument
Documents
```
What I expected is what is seen on the command line above, without the two extra lines and no need to hit RETURN to continue. That is what happens if I type `ls Doc` and hit TAB.

**Comment by xiaq** [comment]
OK, this actually is a feature. Citing https://elv.sh/ref/edit.html#editcompletionsmart-start:

> Starts the completion mode. However, if all the candidates share a non-empty prefix and that prefix starts with the seed, inserts the prefix instead.

There is no special handling for automatically accepting a single candidate; it falls in the general case; when there is only a single candidate, the non-empty prefix is that candidate itself.

However, `edit:completion:smart-start` only inserts the prefix if it also *starts with the seed*. In this case, the seed is `de` (lower case), while the prefix is `Desktop` (with an upper-case D). So the prefix insertion path is not triggered.

---

Consider this case: you have a *substring matcher*, and you have `foo` and `bar` in a directory. Typing `vim o<Tab>` completes to a single candidate `foo`. Since `foo` does not start with the seed `o`, Elvish opens up the completion menu. This is because there is a non-zero chance that you have mistyped and did not mean to complete `foo`. So this is Elvish asking you "are you sure you want to complete `foo`"?

The same can be said for this issue, but the point is a bit weaker. Often we know that `Desktop` starts with a capital `D`, but we are just too lazy to press that Shift key. So when I type `vim de<Tab>`, there is actually a very high chance that I **do** mean to complete `Desktop`.

I think there is a case for `edit:completion:smart-start` to ignore the case or apply smart-case when determining whether prefix completion should be done. It can accept `&smart-case` or `&ignore-case` like the matchers do, and we can default to `&smart-case`.

**Comment by hanche** [comment]
Thanks for the explanation. I don't suppose it would be possible to write elvish code to override the behaviour of `edit:completion:smart-start`? The documentation seems a bit sparse on that point.

**Comment by xiaq** [comment]
No, you can't do that in Elvish code. The Go code needs to be changed.

**Comment by hanche** [comment]
Okay, good to know. I'll be happy to see a fix, but it's not a high priority to me. It's merely a minor irritant.
